### ISSUE TITLE
Re-justify or remove the discussion of `FluentIterable` in "the `ImmutableList.Builder` case"

### ISSUE THREAD

**Issue by cpovirk** [jspecify/jspecify#521]
https://jspecify.dev/docs/spec/#substitution explains why our substitution rules have a special case for null-exclusive type variables. That overall point is still true. But we go on to discuss why we use `MINUS_NULL` instead of `NO_CHANGE`, and we justify it by appealing to `FluentIterable`, whose annotations have subsequently changed with the introduction of `@NonNull`. We should see if the justification still holds up (perhaps with a `FluentIterable`-like type that is null-unmarked) or not, and we should update the spec accordingly.

**Comment by cpovirk** [comment]
Maybe also explain that "the `ImmutableList.Builder` case" has an effect not just for unspecified-nullness type arguments but for _nullable_ type arguments?

In one way, the latter case sounds less important: It can arise only in the case of code that would already have other nullness errors if it were all being checked for nullness.

In another way, it sounds _more_ important: We're generally more comfortable with the idea of leaving types unspecified and letting tools choose to refine them than we are with the idea of leaving types _nullable_ and letting tools choose to refine them. (OK, we actually expect plenty of refining of nullable types, too. But we expect it in the case of "I know that this nullable expression is non-null in this case," not "I know that this nullable _parameter_ is non-null in this case.")

**Comment by netdpb** [comment]
Hmm. By "nullable type arguments" you mean cases like this?

```java
@NullMarked
interface Foo<T> {
  void foo(T t);
}

@NullMarked
// test:not-nullness-subtype:String?:String!:Foo#T
void test(Foo<@Nullable String> foo) {
  foo(null); // not-nullness-subtype or no?
}
```

I continue to think that a smart tool should know that any expression using a symbol whose declaration had a type argument that wasn't a nullness subtype of its type parameter's bound is suspect, certainly if the expression involves that type variable, without the spec making any extra effort to point that out.

**Comment by cpovirk** [comment]
Right, a case like that.

I suspect we agree that the spec and conformance tests should either say that there's a not-nullness-subtype finding there or say that there's not? But I'm not sure what you have in mind when you refer to "making any extra effort." I think we've talked about removing the special case entirely (and a special case is clearly "extra effort" of a sort). Another possibility is to use `NO_CHANGE` instead of `MINUS_NULL` (which wouldn't change the effect on the example you just gave but which might feel simpler), which is more what this issue is about.

Maybe we can add this to the list of things to talk more about in a meeting :)

**Comment by netdpb** [comment]
If this issue is just about changing the non-normative discussion in the spec, then we don't have to debate it here; let's talk about the actual change you want to make (like, in a PR).

If this issue is about changing the normative part of the spec that says that when substituting a type argument for a type parameter, then use `NO_CHANGE` for all the type variables (before any projections) regardless of the operator of the type argument, then the issue title is wrong. 😁

**Comment by netdpb** [comment]
I'll update my last comment to say that if this issue is about changing only the non-normative parts of the spec, then perhaps it shouldn't block the 1.0 release of the spec.

**Comment by cpovirk** [comment]
OK, I have convinced myself that it does still make sense for the rule to use `MINUS_NULL` (as opposed to `NO_CHANGE`). So there is no normative spec change to be made, only an improvement to the non-normative part. Thus:

- This issue doesn't block spec 1.0, as noted above. I'm removing it from the milestone.
- I've mailed https://github.com/jspecify/jspecify/pull/657 to update the example. (But now it looks like I'm going to _remove_ the example instead :))
